### PR TITLE
Prevent duplication org/space naming using leases

### PR DESCRIPTION
**Pull request by kieron-dev**
## Is there a related GitHub Issue?
#285

## What is this change about?
Listing existing names in a validation webhook to prevent duplicate names is inherently racy. The change uses the idea of leases discussed in [this proposal](https://docs.google.com/document/d/1fYIulNhfXlp_y2Ah5EtoKDDX83s6V-WRu162xpziLZ8/edit#) to implement the webhook robustly.

## Does this PR introduce a breaking change?
No

## Acceptance Steps
See repro in issue #285.

### CODE REVIEW

File: controllers/webhooks/workloads/subnamespaceanchor_validation_test.go [hunk 1]
```diff
@@ -324,41 +351,183 @@ var _ = Describe("SubnamespaceanchorValidation", func() {
...
+			When("the org has a space label as well", func() {
+				BeforeEach(func() {
+					anchor.Labels[workloads.SpaceNameLabel] = "my-space"
+				})
+
+				It("allows the deletion anyway", func() {
 					Expect(response.Allowed).To(BeTrue())
 				})
+
+				It("does not attempt to deregister any names", func() {
```
Comment (danail-branekov) [inline controllers/webhooks/workloads/subnamespaceanchor_validation_test.go:506]: We should probably have the `does not attempt to deregister any names` test in the previous context (`the anchor has no org or space labels`)

File: controllers/coordination/name_registry.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hashName(r.entityType, name),
+			Namespace: namespace,
+		},
+	}
+
+	copy := lease.DeepCopy()
+	copy.Spec.HolderIdentity = &unlockedIdentity
+
+	if err := r.client.Patch(ctx, copy, client.MergeFrom(lease)); err != nil {
```
Comment (danail-branekov) [inline controllers/coordination/name_registry.go:110]: @georgethebeatle and I were wondering whether we should allow unlocking names that are not locked, i.e. use the `test-replace` patch operation similar to the one in `TryLockName`. 
On one hand, not doing that makes the unlock operation idempotent, on the other, if you are unlocking an unlocked object, that might be an unexpected state (like when you find your car opened, you would start wondering whether someone else has the keys :))
What do you think?

**Review by danail-branekov — APPROVED** [review]
LGTM, just consider whether my comments make sense. Otherwise, feel free to merge

File: controllers/webhooks/workloads/subnamespaceanchor_validation_test.go (hunk 1 above)
Comment (kieron-dev) [reply controllers/webhooks/workloads/subnamespaceanchor_validation_test.go:506]: Thanks! We've added some more checks for not calling the registry in the failure cases.

File: controllers/coordination/name_registry.go (hunk 2 above)
Comment (kieron-dev) [reply controllers/coordination/name_registry.go:110]: This seems reasonable. We've made the change :)

**Pull request merged**